### PR TITLE
Annotate std/bigint.d and std/internal/math/biguintcore.d to please d…

### PR DESCRIPTION
**Pull request by nordlow**
…lang/dmd#12520

Part of #8076

### CODE REVIEW

File: std/bigint.d [hunk 1]
```diff
@@ -472,13 +472,14 @@ public:
...
      */
-    BigInt opBinary(string op, T)(T y) pure nothrow @safe const
+    BigInt opBinary(string op, T)(T y) pure nothrow @safe const return scope
         if ((op=="+" || op == "*" || op=="-" || op=="|" || op=="&" || op=="^" ||
             op=="/" || op=="%")
             && is (T: BigInt))
     {
         BigInt r = this;
-        return r.opOpAssign!(op)(y);
+        r.opOpAssign!(op)(y);
```
Comment (nordlow) [inline std/bigint.d:481]: I'm gonna see if I can get make this build with this change.

**Comment by nordlow** [comment]
What is the correct parameter qualification of 

```d
    static void divMod(scope return BigUint x, scope return BigUint y,
                       scope out BigUint quotient, scope out BigUint remainder) pure nothrow @safe
    {
        if (y.data.length > x.data.length)
        {
            quotient = 0uL;
            remainder = x;
        }
        else if (y.data.length == 1)
        {
            quotient = divInt(x, y.data[0]);
            remainder = BigUint([modInt(x, y.data[0])]);
        }
        else
        {
            BigDigit[] quot = new BigDigit[x.data.length - y.data.length + 1];
            BigDigit[] rem = new BigDigit[y.data.length];
            divModInternal(quot, rem, x.data, y.data);
            quotient = BigUint(removeLeadingZeros(trustedAssumeUnique(quot)));
            remainder = BigUint(removeLeadingZeros(trustedAssumeUnique(rem)));
        }
    }
```

, @Geod24 ? I can't find any references on assigning return scope parameter to out scope parameters in the DIP-25 and DIP-1000 specs.

**Comment by thewilsonator** [comment]
As far as I understand:
    `return` qualified parameters on a `void` returning function are nonsensical. 
    `out` parameters don't make sense as `scope` in a pure function since they can only be derived from either a pure allocator (e.g. the GC) or from other input parameters. scope on function parameters means that that reference shall not escape the scope of the function.

**Comment by nordlow** [comment]
> As far as I understand:
> `return` qualified parameters on a `void` returning function are nonsensical.
> `out` parameters don't make sense as `scope` in a pure function since they can only be derived from either a pure allocator (e.g. the GC) or from other input parameters. scope on function parameters means that that reference shall not escape the scope of the function.

Ok, thanks. I'll adjust.

**Comment by dkorpel** [comment]
Why does `divMod` need `scope` parameters, is there any existing code calling it with `scope` BigInts?
If you have to make it work, you can do:
```D
// biguintcore.d(977):
remainder = BigUint(x.data.idup);
// biguintcore.d(981):
quotient = divInt(BigUint(x.data.idup), y.data[0]);
```
If you do that, maybe add a comment that the redundant allocation is due to the limitation that `return scope` for out parameters is not supported.

**Comment by nordlow** [comment]
> Why does `divMod` need `scope` parameters, is there any existing code calling it with `scope` BigInts?
> If you have to make it work, you can do:
> 
> ```d
> // biguintcore.d(977):
> remainder = BigUint(x.data.idup);
> // biguintcore.d(981):
> quotient = divInt(BigUint(x.data.idup), y.data[0]);
> ```
> 
> If you do that, maybe add a comment that the redundant allocation is due to the limitation that `return scope` for out parameters is not supported.

This seems like a severe limitation. Has `return scope` for out parameters been considered, @WalterBright?

**Comment by nordlow** [comment]
I've reverted to a the @safe defensive variant 

```
static void divMod(BigUint x, scope BigUint y,
                       out BigUint quotient, out BigUint remainder) pure nothrow @safe
```

that passes locally with dmd on master my machine.

**Comment by nordlow** [comment]
> Why does `divMod` need `scope` parameters, is there any existing code calling it with `scope` BigInts?
> If you have to make it work, you can do:

Thanks for highlighting this. I've reverted to non-x-scope variant 

```
static void divMod(BigUint x, scope BigUint y,
                       out BigUint quotient, out BigUint remainder) pure nothrow @safe
```

that doesn't require any @trusted and that passes locally with dmd master my machine.

**Review by dkorpel — APPROVED** [review]

**Review by WalterBright — APPROVED** [review]

**Pull request merged**